src/Microsoft.Diagnostics.EventFlow.Outputs.EventHub/EventHubOutput.cs

```diff
@@ -83,52 +84,51 @@ public EventHubOutput(IConfiguration configuration, IHealthReporter healthReport
 
             try
             {
-                // Since event hub limits each message/batch to be a certain size, we need to
-                // keep checking the size for exceeds and split into a new batch as needed
+                var groupedEventData = events.GroupBy(e => string.IsNullOrEmpty(outputConfiguration.PartitionKeyProperty) == false && e.TryGetPropertyValue(outputConfiguration.PartitionKeyProperty, out var partionKeyData) ? partionKeyData.ToString() : string.Empty, e => e);
 
-                List<List<MessagingEventData>> batches = new List<List<MessagingEventData>>();
-                int batchByteSize = 0;
+                List<Task> tasks = new List<Task>();
 
-                foreach (EventData eventData in events)
+                foreach (var partitionedEventData in groupedEventData)
                 {
-                    int messageSize;
-                    MessagingEventData messagingEventData = eventData.ToMessagingEventData(out messageSize);
+                    SendBatch(partitionedEventData.Select(e => new Tuple<MessagingEventData, int>(e.ToMessagingEventData(out var messageSize), messageSize)).ToList());
 
-                    // If we don't have a batch yet, or the addition of this message will exceed the limit for this batch, then
-                    // start a new batch.
-                    if (batches.Count == 0 ||
-                        batchByteSize + messageSize > EventHubMessageSizeLimit)
+                    void SendBatch(List<Tuple<MessagingEventData, int>> batch)
                     {
-                        batches.Add(new List<MessagingEventData>());
-                        batchByteSize = 0;
+                        // Since event hub limits each message/batch to be a certain size, we need to
+                        // keep checking the size in bytes of the batch and recursively keep splitting into two batches as needed
+                        if (batch.Count >= 2 && batch.Sum(b => b.Item2) > EventHubMessageSizeLimit)
+                        {
+                            var indexMiddle = batch.Count / 2;
+                            SendBatch(batch.Take(indexMiddle).ToList());
```

Comment:
Now with the extensive use of Linq you are re-calculating sizes of candidate batches and iterating over the same elements many times. I think it is better to just keep taking elements from the partition until you exceed the batch size limit, send the batch, start a new one and keep going until you run out of partition elements. It is both easier to understand and you consider each partition element just once.